### PR TITLE
Add a new Location.Create overload for external file location with a different mappedLineSpan

### LINKED ISSUE #64236: Add a new Location.Create overload for external file location with a different mappedLineSpan
## Background and Motivation
TypeScript would like to create `ExternalFileLocation` for TS diagnostics reported on their secondary projection buffers, such that the diagnostic location has a different mapped location pointing to the primary buffer (separate file and line span), for display purposes, such as in the error list. The core story behind their request is as below:

> The story is that currently projection script blocks in HTML and ASPX have the same URI as the original file, causing problems with LSP scenarios since the editor expects unique URIs for every buffer. This produces all kinds of buffer/snapshot errors, including diagnostic squiggles being shifted, exceptions from the LSP editor code, etc. Todd helped me get a fix for this that involves using a [different name for these projection buffers](https://devdiv.visualstudio.com/DevDiv/_git/WebTools/commit/587323f525ca72627c26d94799de5aa717481b06?refName=refs%2Fheads%2Fdev%2Ftoddgrun%2FAddContainedBufferTextDocumentName3), but then I had to [update TypeScript-VS](https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_git/TypeScript-VS/pullrequest/423511) because all of our script block handling code has always assumed that projection buffers use the same file name as the original documents.

> And so now the update in TypeScript-VS is almost ready to go except for this issue with diagnostics, since Roslyn won't allow me to display diagnostics with the primary/mapped buffer's file name (since the workspace doesn't have a document with that name, it has been using the secondary buffer name).

## Proposed API
Currently, `Location.Create` public API for creating external file locations only supports providing a single lineSpan, which is used as both the actual line span and the mapped line span.

https://github.com/dotnet/roslyn/blob/620a86b61272b794670bbbf5a425bfa7ec59ce91/src/Compilers/Core/Portable/Diagnostic/Location.cs#L168-L179

https://github.com/dotnet/roslyn/blob/0f83d53179cd936842c743ad5dd48bc41cbbfb80/src/Compilers/Core/Portable/Diagnostic/ExternalFileLocation.cs#L35-L43

The public API request is to add a new `Location.Create` overload that supports mapped file + line span, either one of the below public APIs would suffice.

### Proposal 1
Draft PR for this proposal: #64240
```diff
namespace Microsoft.CodeAnalysis
{
     public abstract class Location
     {
+        public static Location Create(string filePath, TextSpan textSpan, LinePositionSpan lineSpan, string mappedFilePath, LinePositionSpan mappedLineSpan);
     }
```

### Proposal 2
```diff
namespace Microsoft.CodeAnalysis
{
     public abstract class Location
     {
+        public static Location Create(TextSpan textSpan, FileLinePositionSpan lineFileSpan, FileLinePositionSpan mappedFileLineSpan);
     }
```

### PR DESCRIPTION
**Pull request by mavasani**
Closes #64236

Proposal details in https://github.com/dotnet/roslyn/issues/64236#issue-1383197448

### CODE REVIEW

**Comment by mavasani** [comment]
ping @dotnet/roslyn-compiler for reviews.

FYI: This is blocking TypeScript team from moving to a new model of how they report diagnostics with primary/secondary buffer.

**Review by 333fred — APPROVED** [review]

**Comment by MariaSolOs** [comment]
> ping @dotnet/roslyn-compiler for reviews.
> 
> FYI: This is blocking TypeScript team from moving to a new model of how they report diagnostics with primary/secondary buffer.

Bumping this. This represents a major change in our script block management and the sooner this change gets merged, the more time we'll have to test this carefully in 17.5.

File: src/Compilers/Core/Portable/Diagnostic/ExternalFileLocation.cs [hunk 1]
```diff
@@ -14,12 +14,13 @@ namespace Microsoft.CodeAnalysis
...
         private readonly TextSpan _sourceSpan;
-        private readonly FileLinePositionSpan _lineSpan;
+        private readonly FileLinePositionSpan _lineSpan, _mappedLineSpan;
 
-        internal ExternalFileLocation(string filePath, TextSpan sourceSpan, LinePositionSpan lineSpan)
+        internal ExternalFileLocation(string filePath, TextSpan sourceSpan, LinePositionSpan lineSpan, string mappedFilePath, LinePositionSpan mappedLineSpan)
         {
             _sourceSpan = sourceSpan;
             _lineSpan = new FileLinePositionSpan(filePath, lineSpan);
+            _mappedLineSpan = new FileLinePositionSpan(mappedFilePath, mappedLineSpan, hasMappedPath: !PathUtilities.PathsEqual(filePath, mappedFilePath));
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/Diagnostic/ExternalFileLocation.cs:23]: is the hasMappedPath portion of this tested?

Comment (MariaSolOs) [reply src/Compilers/Core/Portable/Diagnostic/ExternalFileLocation.cs:23]: There are some checks in the test added by @mavasani:
https://github.com/dotnet/roslyn/blob/114f08f3bf747f8b8dedfa5addf11b197687d383/src/Compilers/CSharp/Test/Syntax/Diagnostics/LocationsTests.cs#L529-L547

Comment (MariaSolOs) [reply src/Compilers/Core/Portable/Diagnostic/ExternalFileLocation.cs:23]: Or were you asking for a deeper testing @CyrusNajmabadi?

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler can we get a second pair of eyes on this pr?

File: src/Compilers/Core/Portable/Diagnostic/ExternalFileLocation.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/Diagnostic/ExternalFileLocation.cs:23]: nope.  that's answers the question.

Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Diagnostic/ExternalFileLocation.cs:23]: >!PathUtilities.PathsEqual(filePath, mappedFilePath) [](http://example.com/codeflow?start=102&length=51)

What if ```lineSpan``` != ```mappedLineSpan```
I do not think we should use any heuristic here. Consider having two separate constructors here. One that takes mapped information and one that doesn't. The one that takes mapped information always passes true here.

File: src/Compilers/Core/Portable/Diagnostic/Location.cs [hunk 2]
```diff
@@ -175,7 +175,25 @@ public static Location Create(string filePath, TextSpan textSpan, LinePositionSp
                 throw new ArgumentNullException(nameof(filePath));
             }
 
-            return new ExternalFileLocation(filePath, textSpan, lineSpan);
+            return Create(filePath, textSpan, lineSpan, filePath, lineSpan);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Diagnostic/Location.cs:178]: >return Create(filePath, textSpan, lineSpan, filePath, lineSpan); [](http://example.com/codeflow?start=12&length=64)

Consider keeping the constructor and continue using it here.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 4)

**Comment by MariaSolOs** [comment]
Thanks @AlekseyTs for the review! I applied your feedback in https://github.com/mavasani/roslyn/pull/3, (I've been piggy-backing on @mavasani's branch).

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 6). Please squash commit in the process of merging.

**Pull request merged**